File: subprojects/core/src/main/java/hu/bme/mit/theta/core/type/abstracttype/RemExpr.java
package hu.bme.mit.theta.core.type.abstracttype;

import hu.bme.mit.theta.core.type.BinaryExpr;
import hu.bme.mit.theta.core.type.Expr;

import static hu.bme.mit.theta.core.utils.TypeUtils.cast;

public abstract class RemExpr<ExprType extends Divisible<ExprType>> extends BinaryExpr<ExprType, ExprType> {

    protected RemExpr(final Expr<ExprType> leftOp, final Expr<ExprType> rightOp) {
        super(leftOp, rightOp);
    }

    public static <ExprType extends Divisible<ExprType>> RemExpr<?> create2(final Expr<?> leftOp, final Expr<?> rightOp) {
        @SuppressWarnings("unchecked") final ExprType type = (ExprType) leftOp.getType();
        final Expr<ExprType> newLeftOp = cast(leftOp, type);
        final Expr<ExprType> newRightOp = cast(rightOp, type);
        return type.Rem(newLeftOp, newRightOp);
    }
}

File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/abstracttype/RemExpr.kt
package hu.bme.mit.theta.core.type.abstracttype

import hu.bme.mit.theta.core.type.BinaryExpr
import hu.bme.mit.theta.core.type.Expr
import hu.bme.mit.theta.core.utils.TypeUtils.cast
import kotlinx.serialization.Serializable

/**
 * Abstract base class for remainder expressions over divisible types.
 */
@Serializable
abstract class RemExpr<ExprType : Divisible<ExprType>> : BinaryExpr<ExprType, ExprType>() {

    companion object {
        @JvmStatic
        fun <ExprType : Divisible<ExprType>> create2(
            leftOp: Expr<*>,
            rightOp: Expr<*>
        ): RemExpr<*> {
            @Suppress("UNCHECKED_CAST")
            val type = leftOp.type as ExprType
            val newLeftOp = cast(leftOp, type)
            val newRightOp = cast(rightOp, type)
            return type.Rem(newLeftOp, newRightOp)
        }
    }
}

